src/utility/HCISharedMemTransport.cpp

```diff
@@ -0,0 +1,837 @@
+/*
+  This file is part of the STM32duinoBLE library.
+  Copyright (c) 2019 STMicroelectronics. All rights reserved.
+
+  This library is free software; you can redistribute it and/or
+  modify it under the terms of the GNU Lesser General Public
+  License as published by the Free Software Foundation; either
+  version 2.1 of the License, or (at your option) any later version.
+
+  This library is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+  Lesser General Public License for more details.
+
+  You should have received a copy of the GNU Lesser General Public
+  License along with this library; if not, write to the Free Software
+  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+*/
+
+#include "HCISharedMemTransport.h"
+#include "STM32Cube_FW/hw.h"
+
+/* global var used as semaphore to control incoming events */
+volatile bool sys_event; /* true : M0 core is already up and running */
+volatile bool sys_resp;
+volatile bool acl_data; /* true : data acknowledged by the BLE */
+
+#define CONFIG_DEBUG
+
+#define HW_IPCC_Tx_Handler IPCC_C1_TX_IRQHandler
+#define HW_IPCC_Rx_Handler IPCC_C1_RX_IRQHandler
+extern "C" { void HW_IPCC_Tx_Handler(void);}
+extern "C" { void HW_IPCC_Rx_Handler(void);}
+
+/* Private variables ---------------------------------------------------------*/
+PLACE_IN_SECTION("MB_MEM1") ALIGN(4) static TL_CmdPacket_t BleCmdBuffer;
+PLACE_IN_SECTION("MB_MEM2") ALIGN(4) static uint8_t EvtPool[POOL_SIZE];
+PLACE_IN_SECTION("MB_MEM2") ALIGN(4) static TL_CmdPacket_t SystemCmdBuffer;
+PLACE_IN_SECTION("MB_MEM2") ALIGN(4) static uint8_t
+	SystemSpareEvtBuffer[sizeof(TL_PacketHeader_t) + TL_EVT_HDR_SIZE + 255];
+PLACE_IN_SECTION("MB_MEM2") ALIGN(4) static uint8_t
+	BleSpareEvtBuffer[sizeof(TL_PacketHeader_t) + TL_EVT_HDR_SIZE + 255];
+PLACE_IN_SECTION("MB_MEM2") ALIGN(4) static uint8_t
+	HciAclDataBuffer[sizeof(TL_PacketHeader_t) + 5 + 251];
+
+HCISharedMemTransportClass::HCISharedMemTransportClass(BLEChip_t ble_chip) :
+  _ble_chip(ble_chip)
+{
+  _read_index = 0;
+  _write_index = 0;
+  _write_index_initial = 0;
+
+  sys_event = false;
+  sys_resp = false;
+  acl_data = false;
+}
+
+static void evt_received(TL_EvtPacket_t *hcievt)
+{
+    uint16_t len = 0;
+
+    // We need to memcpy the data before passing to higher layers.
+    switch (hcievt->evtserial.type) {
+        case TL_BLEEVT_PKT_TYPE:
+            len = hcievt->evtserial.evt.plen + TL_EVT_HDR_SIZE;
+            //TODO: on_data_received((uint8_t *)&hcievt->evtserial, len);
+            break;
+        case TL_ACL_DATA_PKT_TYPE: {
+            TL_AclDataSerial_t *acl = &(((TL_AclDataPacket_t *)hcievt)->AclDataSerial);
```

Comment:
same here - BTW maybe Arduino BLE handles ACL and BLEEVT in the same way ??

Reply:
format of each packet differs